bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtoEncoding.java

```diff
@@ -309,6 +381,25 @@ public Object decode(ByteBuf buffer)
 
                 return new BookieProtocol.ReadResponse(
                         version, rc, ledgerId, entryId, buffer.retainedSlice());
+            case BookieProtocol.BATCH_READ_ENTRY:
+                rc = buffer.readInt();
+                ledgerId = buffer.readLong();
+                entryId = buffer.readLong();
+                long requestId = buffer.readLong();
+                ByteBufList data = null;
+                while (buffer.readableBytes() > 0) {
+                    int entrySize = buffer.readInt();
+                    int entryPos = buffer.readerIndex();
+                    if (data == null) {
+                        data = ByteBufList.get(buffer.retainedSlice(entryPos, entrySize));
+                        buffer.readerIndex(entryPos + entrySize);
+                    } else {
+                        data.add(buffer.retainedSlice(entryPos, entrySize));
+                        buffer.readerIndex(entryPos + entrySize);
+                    }
+                }
+                return new BookieProtocol.BatchedReadResponse(version, rc, ledgerId, entryId, requestId, data == null
```

Comment:
Why not pass null directly instead of pass an empty ByteBufList when data is null?

Reply:
Using empty ByteBufList can avoid NPE. And I check the existing code; it return null directly when the `rc != Ok`. I will follow the code style.

Reply:
I checked it again; we would better use empty ByteBufList, or [BatchedReadResponse#touch()](https://github.com/apache/bookkeeper/blob/0b0504fed09fccfb4e71c19ea29d97b8bcb6a416/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtocol.java#L603)) will throw NPE.